src/v/cluster/archival/archiver_scheduler_impl.cc

```diff
@@ -0,0 +1,152 @@
+                   .cloud_storage_upload_loop_max_backoff_ms.bind()) {}
+
+template<class Clock>
+ss::future<result<upload_resource_quota>>
+archiver_scheduler<Clock>::maybe_suspend_upload(
+  upload_resource_usage<Clock> arg) noexcept {
+    auto holder = _gate.hold();
+    auto it = _partitions.find(arg.ntp);
+    if (it == _partitions.end()) {
+        // create_ntp_state not called for ntp
+        vlog(archival_log.error, "Unexpected NTP {}", arg.ntp);
+        co_return error_outcome::unexpected_failure;
+    }
+    auto& v = it->second;
+    auto ntp_holder = v->gate.hold();
+    if (arg.errc) {
+        // The error has occurred. We need to apply exponential backoff
+        // to avoid consuming requests.
+        v->backoff = v->backoff.value_or(0ms)
+                     + v->backoff.value_or(_initial_backoff());
+        v->backoff = std::clamp(v->backoff.value(), 0ms, _max_backoff());
+
+        auto sleep_time = v->backoff.value();
+        auto now = Clock::now();
+        if (
+          arg.next_housekeeping.has_value()
+          && arg.next_housekeeping.value() > now) {
+            // In case if the ntp_archiver reported its next
+            // housekeeping timestamp or next manifest reupload timestamp
+            // we can override the sleep_time to wake up early.
+            // This will not affect the backoff time (which will affect
+            // next sleep). It will only replace current sleep time.
+            auto dt = arg.next_housekeeping.value() - now;
+            if (dt < sleep_time) {
+                sleep_time
+                  = std::chrono::duration_cast<std::chrono::milliseconds>(dt);
+            }
+        }
+        if (
+          arg.next_manifest_upload.has_value()
+          && arg.next_manifest_upload.value() > now) {
+            auto dt = arg.next_manifest_upload.value() - now;
+            if (dt < sleep_time) {
+                sleep_time
+                  = std::chrono::duration_cast<std::chrono::milliseconds>(dt);
+            }
+        }
+
+        // Backoff is applied in case of any error including no-data error.
+        try {
+            co_await ss::sleep_abortable<Clock>(
+              sleep_time, arg.archiver_rtc.get().root_abort_source());
+        } catch (...) {
+            auto er = std::current_exception();
+            if (ssx::is_shutdown_exception(er)) {
+                co_return error_outcome::shutting_down;
+            }
+            vlog(archival_log.error, "Unexpected sleep error {}", er);
+        }
+    } else {
+        // No error so we can proceed to next upload immediately
+        // if rate limits are not reached.
+        v->backoff.reset();
+    }
+    auto puts = arg.put_requests_used > 0 ? arg.put_requests_used : 1;
+    try {
+        co_await _put_requests.throttle(
+          puts, arg.archiver_rtc.get().root_abort_source());
+        co_await _shard_tput_limit.throttle(
+          arg.uploaded_bytes, arg.archiver_rtc.get().root_abort_source());
```

Comment:
Just curious, would it make sense to perform these waits in parallel? They seem independent, so performing them serially seems a bit odd